src/Solvers/multigrid_solver.jl

```diff
@@ -0,0 +1,168 @@
+using Oceananigans.Architectures: architecture
+using Oceananigans.Grids: interior_parent_indices
+using Statistics: norm, dot
+using LinearAlgebra
+using AlgebraicMultigrid: _solve!, init, RugeStubenAMG
+using Oceananigans.Operators: volume, Δyᶠᶜᵃ, Δyᶜᶠᵃ, Δyᶜᶜᵃ, Δxᶠᶜᵃ, Δxᶜᶠᵃ, Δxᶜᶜᵃ, Δyᵃᶜᵃ, Δxᶜᵃᵃ, Δzᵃᵃᶠ, Δzᵃᵃᶜ, ∇²ᶜᶜᶜ
+
+import Oceananigans.Architectures: architecture
+
+mutable struct MultigridSolver{A, G, L, T, M, F}
+               architecture :: A
+                       grid :: G
+            linear_operator :: L
+                     abstol :: T
+                     reltol :: T
+                    maxiter :: Int
+              amg_algorithm :: M
+                    x_array :: F
+                    b_array :: F
+end
+
+architecture(solver::MultigridSolver) = solver.architecture
+    
+"""
+    MultigridSolver(grid,
+                    linear_operation!, 
+                    args...;
+                    maximum_iterations = 100,
+                    tolerance = 1e-13,
+                    amg_algorithm = RugeStubenAMG(),
+                    )
+
+Returns a MultigridSolver that solves the linear equation
+``A x = b`` using a multigrid method, where `A * x` is
+determined by `linear_operation!`
+
+`linear_operation!` is a function with signature `linear_operation!(Ax, x, args...)` 
+that calculates `A * x` for given `x` and stores the result in `Ax`.
+
+
+The solver is used by calling
+
+```
+solve!(x, solver::MultigridSolver, b; kwargs...)
+```
+
+for `solver`, right-hand side `b`, solution `x`, and optional keyword arguments `kwargs...`.
+
+Arguments
+=========
+
+* `template_field`: Dummy field that is the same type and size as `x` and `b`, which
+                    is used to infer the `architecture`, `grid`, and to create work arrays
+                    that are used internally by the solver.
+
+* `maxiter`: Maximum number of iterations the solver may perform before exiting.
+
+* `reltol, abstol`: Relative and absolute tolerance for convergence of the algorithm.
+                    The iteration stops when `norm(A * x - b) < tolerance`.
+"""
+function MultigridSolver(linear_operation!::Function,
+                         args...;
+                         template_field::AbstractField,
+                         maxiter = prod(size(template_field)),
+                         reltol = sqrt(eps(eltype(template_field.grid))),
+                         abstol = 0reltol,
+                         amg_algorithm = RugeStubenAMG(),
+                         )
+
+    arch = architecture(template_field)
+    grid = template_field.grid
+
+    matrix = create_matrix(template_field, linear_operation!, args...)
+
+    Nx, Ny, Nz = size(template_field)
+
+    FT = eltype(grid)
```

Comment:
```suggestion

    matrix = create_matrix(template_field, linear_operation!, args...)

    Nx, Ny, Nz = size(template_field)

    FT = eltype(template_field.grid)
```

Reply:
We also need to change line 83 but I still think it's nicer. We could instead store the field instead of the grid in the constructor? I think it's only used for the `show` method

Reply:
I think I'll keep it as `grid` for now seeing as that agrees with the PCG solver

Reply:
you mean store the template field? I guess so...
do we even need to store that?

Reply:
Indeed, it just might have saved me typing `.grid` :)